components/src/dynamo/profiler/tests/unit/test_helpers_rapid.py

```
@@ -184,6 +184,33 @@ def fake_generate(params, backend, use_dynamo_generator=False):
+                },
+            ),
+            patch("dynamo.profiler.rapid.generate_backend_artifacts") as mock_generate,
+        ):
+            result = _run_naive_fallback(dgdr, "Qwen/Qwen3-32B", 2, "l40s", "vllm")
+
+        assert result["error"] == "insufficient_gpus_for_model"
+        assert result["dgd_config"] is None
+        assert result["best_config_df"].empty
+
+    # Critical: downstream generation must NOT happen
+    mock_generate.assert_not_called()
```

Ruff (0.15.12)

[error] Undefined name `mock_generate`

(F821)